(function($){    $.fn.carbCalc= function() {        var self=this,    	calcFunc, getParms, showResults,    	fields,    	totalNotCarbs="<tr><td class='label'>Total:</td><td><span id='totalNotCarbs' class='result'></span></td></tr>",    	totalCarbs="<div class='label'> Total Carbs: <span id='totalCarbs' class='result'></span></div>",    	dryMatter="<div class='label'> Dry Matter: <span id='dryMatter' class='result'></span></div>",    	dryCarbs="<div class='label'> Dry Carbs: <span id='dryCarbs'  class='result'></span></div>",    	fieldTemplate="<tr><td class='label'><!label!></td<td><input type='text' name='<!fieldName!>'></td></tr>",    	outString='',    	tmpString='',    	intro="<div class='intro'>Reading the cat food label, 'Guaranteed Analysis', and enter the values in the boxes below.</div>",    	title="<div class='title'>Cat Carb Calculator</div>",    	moral="<div class='moral'>Doctor Craig says, \"There is no magic number, but most cats do best when about 10% of their diet is carbs. If a cat is overweight, the dry carb percentage should be even less.\"</div>";	calcFunc=function(){				var values=getParms(),				totalNotCarbs=0,				outObj={};								for (var i in values){					totalNotCarbs+=values[i];				}								outObj.totalNotCarbs=totalNotCarbs;				outObj.totalCarbs=100-totalNotCarbs;				outObj.dryMatter=100-values.moisture;				outObj.dryCarbs=((outObj.totalCarbs/outObj.dryMatter).toFixed(2)*100).toString();								return outObj;			};				getParms=function(){		var elements=$("input[name], textarea[name], select[name]", self),			outObj={};				$.each(elements, function(inx, item){			outObj[item.name]=$(item).val()*1.0;		});		return outObj;	};		showResults=function(){		var results=calcFunc();		$('#totalNotCarbs').text(results.totalNotCarbs+'%');		$('#totalCarbs').text(results.totalCarbs+'%');		$('#dryMatter').text(results.dryMatter+'%');		var dryCarbsDomObj=$('#dryCarbs').text(results.dryCarbs+'%');				if (results.dryCarbs>10){			dryCarbsDomObj.css('color','red');		}		else{			dryCarbsDomObj.css('color','green');		}	}			fields=[    			{label:'Protein %', fieldName:'protein'},		{label:'Fat %', fieldName:'fat'},		{label:'Fiber %', fieldName:'fiber'},		{label:'Moisture %', fieldName:'moisture'},		{label:'Ash %', fieldName:'ash'}	];		for (var i=0, len=fields.length; i<len; i++){		tmpString=fieldTemplate;		tmpString=tmpString.replace('<!label!>', fields[i].label);		tmpString=tmpString.replace('<!fieldName!>', fields[i].fieldName);		outString+=tmpString;		tmpString='';	};		outString="<div class='panel'>"+title+intro+"<table>"+outString+totalNotCarbs+"</table>"+totalCarbs+dryMatter+dryCarbs+moral+"</div>";		$('.result').css({		'fontSize':'14pt',		'color':'green',		'fontWeight':'bold'	});      this.append(outString);  	$('.result', this).css({		'fontSize':'14pt',		'color':'green',		'fontWeight':'bold'	});		$('.label', this).css({		'fontSize':'12pt',		'color':'blue',		'fontWeight':'normal'	});		$('.intro', this).css({		'fontSize':'10pt',		'color':'green',		'fontWeight':'normal',		marginTop:'10px',		marginBottom:'10px',		borderBottom:'1pt solid gray',		paddingBottom:'10px'	});		$('.title', this).css({		'fontSize':'12pt',		'color':'black',		'fontWeight':'bold',		marginBottom:'10px',		borderBottom:'1pt solid gray',		paddingBottom:'10px',		textAlign:'center'	});		$('.moral', this).css({		'fontSize':'10pt',		'color':'gray',		'fontWeight':'normal',		marginTop:'10px',		marginBottom:'10px',		borderTop:'1pt solid gray',		paddingTop:'10px'	});		this.css({		width:'250px',		padding:'10px',		border:'1pt solid gray',		margin:'5px',		fontFamily:'sans-serif'	});    $("input[name], textarea[name], select[name]", self).change(showResults);    $($('input')[0]).focus();            };    })(jQuery);